src/tests/benchmark/benchmark_controller.h

```diff
@@ -0,0 +1,167 @@
+                                   std::chrono::steady_clock::now() - start)
+                                   .count();
+            results_.emplace(test_name, elapsed);
+            std::cout << "** " << test_name << ": " << elapsed / 1000 << "s: ";
+            // Assume that benchmark cannot run 0ms
+            if (prev_results_.contains(test_name) && prev_results_[test_name] != 0) {
+                auto prev_res = prev_results_[test_name];
+                auto overhead = static_cast<double>(elapsed - prev_res) / prev_res * 100;
+                std::cout << (overhead <= threshold ? "SUCCESS" : "FAIL") << ": "
+                          << std::setprecision(3) << std::abs(overhead) << "% "
+                          << (overhead > 0 ? "slower" : "faster") << " than previous run **\n";
+                return (overhead <= threshold);
+            }
+            std::cout << "WARNING : hasn't been ran before **\n";
+            return true;
+        };
+    }
+
+public:
+    BenchmarkController(BenchmarkController&) = delete;
+    BenchmarkController(BenchmarkController&&) = default;
+    BenchmarkController& operator=(BenchmarkController&) = delete;
+    BenchmarkController& operator=(BenchmarkController&&) = default;
+    ~BenchmarkController() = default;
+
+    /// @brief Get an instance of @c BenchmarkController.
+    static BenchmarkController& Instance() {
+        static BenchmarkController instance;
+        return instance;
+    }
+
+    /// @brief Load results of previous run.
+    /// @note It's unlikely that you'll need to call this method.
+    void LoadPreviousResults(std::unordered_map<std::string, long long>&& prev_res) {
+        prev_results_ = std::move(prev_res);
+    }
+
+    /// @brief Register a complex test to be executed at @c RunAllTests call.
+    /// @tparam Test any callable, which returns @a truthy @a value when test passed,
+    /// @a falsy otherwise. See implementation of @c WrapTest as an example.
+    /// @note Use this method only if you really need precise control on which part of test
+    /// is measured. <b>If you don't know which method to use to register your test, most likely
+    /// it's @c RegisterTest(), not this one.</b>
+    template <typename Test>
+    void RegisterCustomTest(Test&& test) {
+        tests_.push_back(std::forward<Test>(test));
+    }
+
+    /// @brief Register test to be executed at @c RunAllTests call.
+    /// @tparam Body any callable.
+    /// @param test_body the whole test.
+    /// @param test_name a test name to be displayed. Must be unique. A good test name includes
+    /// algorithm name and dataset name (or size).
+    /// @param threshold execution time threshold, per cent.
+    /// @note If you need more precise control on which part of test is measured, see @c
+    /// RegisterCustomTest.
+    template <typename Body>
+    void RegisterTest(Body&& test_body, std::string const& test_name,
```

Comment:
```suggestion
    void WrapAndRegisterTest(Body&& test_body, std::string const& test_name,
```
it is not entirely clear what the method does and how it differs from custom